google/cloud/bigquery/dataset.py

```diff
@@ -166,22 +166,24 @@ def from_string(
                 standard SQL format.
         """
         output_dataset_id = dataset_id
-        output_project_id = default_project
         parts = _helpers._split_id(dataset_id)
 
-        if len(parts) == 1 and not default_project:
-            raise ValueError(
-                "When default_project is not set, dataset_id must be a "
-                "fully-qualified dataset ID in standard SQL format, "
-                'e.g., "project.dataset_id" got {}'.format(dataset_id)
-            )
+        if len(parts) == 1:
+            if default_project is not None:
+                output_project_id = default_project
+            else:
+                raise ValueError(
+                    "When default_project is not set, dataset_id must be a "
+                    "fully-qualified dataset ID in standard SQL format, "
+                    'e.g., "project.dataset_id" got {}'.format(dataset_id)
+                )
         elif len(parts) == 2:
             output_project_id, output_dataset_id = parts
-        elif len(parts) > 2:
+        else:
```

Comment:
Here `else:` covers the same cases as `elif len(parts) > 2:`

Parameter `dataset_id` is a string, because of that list `parts` will always have at least 1 item.
By using `else` we can reassure `pytest-cov` that everything is ok.